### PR TITLE
Allow ClientCore.assign to accept const params

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 1]
```diff
@@ -222,6 +272,31 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
...
+            Params:
+                params = user-specified request parameters to store
+                serialize_buffer = intermediary serialization buffer
+
+        ***********************************************************************/
+
+        public void setUserSpecifiedParams ( Const!(UserSpecifiedParams) params,
+            ref ubyte[] serialize_buffer )
+        {
+            enableStomping(serialize_buffer);
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d]: I think you can save on `enableStomping` call here, as the only time you change the length of it is below.

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: Ah yeah, right. Updated.

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
Blocked. Adding a second commit to convert the test client to use this.

**Comment by gavin-norman-sociomantic** [comment]
Updated. Unblocked.

**Comment by gavin-norman-sociomantic** [comment]
Added missing release note about deprecation.

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 2]
```diff
@@ -222,6 +272,30 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
...
+                params = user-specified request parameters to store
+                serialize_buffer = intermediary serialization buffer
+
+        ***********************************************************************/
+
+        public void setUserSpecifiedParams ( Const!(UserSpecifiedParams) params,
+            ref ubyte[] serialize_buffer )
+        {
+            auto serialized = Serializer.serialize(params, serialize_buffer);
+            enableStomping(serialized);
```
Comment (mihails-strasuns-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d:294]: It is a bit weird that `enableStomping` is required here, `serialize_buffer` is not resized in any other place, right?

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 3]
```diff
@@ -178,7 +178,54 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
...
         public Args args;
 
-        public SerializableReferenceType!(Notifier) notifier;
+        /***********************************************************************
+
+            Note: this struct is essentially deprecated. It is only needed in
+            order to maintain the public API of UserSpecifiedParams (specifically
+            the ability to call `params.notifier.set()`). In the next major
+            release, the `serialized_notifier` can be moved into
+            UserSpecifiedParams and this wrapper struct removed.
```
Comment (mihails-strasuns-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d:187]: How about some `TODO / FIXME` ? :)

**Review by mihails-strasuns-sociomantic — APPROVED** [review]

File: src/swarm/neo/client/mixins/RequestCore.d (hunk 3 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:187]: I used the keyword "deprecated" deliberately. We search for that way more often than "TODO" or "FIXME" ;)

Comment (mihails-strasuns-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:187]: Wise :)

File: src/swarm/neo/client/mixins/RequestCore.d (hunk 2 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:294]: Erm, this looks to me like utterly standard usage. 1. resize array (potentially making it shorter), 2. call `enableStomping`...?

Comment (mihails-strasuns-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:294]: It shouldn't trigger if there is only one slice for the given root but I will simply assume the compiler is not as smart and won't investigate :)

**Pull request merged**